src/vcpkg/binarycaching.cpp

```diff
@@ -881,70 +870,73 @@ namespace
+            {
+                auto maybe_json = Json::parse_object(*p);
+                if (auto json = maybe_json.get())
+                {
+                    auto cache_id = json->get("cacheId");
+                    if (cache_id && cache_id->is_integer())
+                    {
+                        return cache_id->integer(VCPKG_LINE_INFO);
+                    }
+                }
+            }
+            return {};
         }
 
         size_t push_success(const BinaryPackageWriteInfo& request, MessageSink&) override
         {
             if (!request.zip_path) return 0;
+
             const auto& zip_path = *request.zip_path.get();
             const ElapsedTimer timer;
             const auto& abi = request.package_abi;
 
-            int64_t cache_size;
-            {
-                auto archive = m_fs.open_for_read(zip_path, VCPKG_LINE_INFO);
-                archive.try_seek_to(0, SEEK_END);
-                cache_size = archive.tell();
-            }
-
             size_t upload_count = 0;
-            if (auto cacheId = reserve_cache_entry(abi, cache_size))
+            auto cache_size = m_fs.file_size(zip_path, VCPKG_LINE_INFO);
+
+            if (auto cacheId = reserve_cache_entry(request.spec.name(), abi, cache_size))
             {
-                std::vector<std::string> headers{
+                std::vector<std::string> custom_headers{
                     m_token_header,
                     m_accept_header.to_string(),
                     "Content-Type: application/octet-stream",
                     "Content-Range: bytes 0-" + std::to_string(cache_size) + "/*",
                 };
                 auto url = m_url + "/" + std::to_string(*cacheId.get());
-                if (put_file(m_fs, url, {}, headers, zip_path, "PATCH"))
+
+                if (put_file(m_fs, url, {}, custom_headers, zip_path, "PATCH"))
                 {
                     Json::Object commit;
                     commit.insert("size", std::to_string(cache_size));
-                    auto cmd = command().string_arg(url).string_arg("-d").string_arg(stringify(commit));
-
-                    auto res = cmd_execute_and_capture_output(cmd);
-                    if (res.has_value() && !res.get()->exit_code)
+                    auto res = invoke_http_request("POST",
+                                                   std::vector<std::string>{m_accept_header.to_string(),
+                                                                            m_content_type_header.to_string(),
+                                                                            m_token_header},
+                                                   url,
+                                                   stringify(commit));
+                    if (res.get())
                     {
                         ++upload_count;
                     }
+                    else
+                    {
+                        msg::println_error(res.error());
```

Comment:
```suggestion
                        msg::println(res.error());
```

`invoke_http_request` already returns errors by calling `format_error` so they already have the `error:` part prepended.